src/server.cc

```diff
@@ -217,20 +217,23 @@ int Server::startupInterface(const std::string& iface, in_port_t inPort)
         iface.empty() ? "<unset>" : iface, inPort == 0 ? "<unset>" : fmt::to_string(inPort));
     const char* ifName = iface.empty() ? nullptr : iface.c_str();
 
+    std::size_t attempt;
     int ret = UPNP_E_INIT_FAILED;
-    for (std::size_t attempt = 0; ret != UPNP_E_SUCCESS; attempt++) {
+    for (attempt = 0; attempt < 4; attempt++) {
         ret = UpnpInit2(ifName, inPort);
-        if (ret != UPNP_E_SUCCESS) {
-            if (attempt > 3) {
-                throw UpnpException(ret, fmt::format("run: UpnpInit failed with {} {}", ifName, inPort));
-            }
+        if (ret == UPNP_E_SUCCESS)
+            break;
+
 #ifdef UPNP_HAVE_TOOLS
-            log_warning("UPnP Init {}:{} failed: {} ({}). Retrying in {} seconds...", ifName, inPort, UpnpGetErrorMessage(ret), ret, attempt + 1);
+        log_warning("UPnP Init {}:{} failed: {} ({}). Retrying in {} seconds...", ifName, inPort, UpnpGetErrorMessage(ret), ret, attempt + 1);
 #else
-            log_warning("UPnP Init {}:{} failed: ({}). Retrying in {} seconds...", ifName, inPort, ret, attempt + 1);
+        log_warning("UPnP Init {}:{} failed: ({}). Retrying in {} seconds...", ifName, inPort, ret, attempt + 1);
 #endif
-            std::this_thread::sleep_for(std::chrono::seconds(attempt + 1));
-        }
+        std::this_thread::sleep_for(std::chrono::seconds(attempt + 1));
+    }
+
+    if (attempt > 3) {
```

Comment:
It's clearer if you check `ret != sucess` here.

Reply:
I think this is what you mean.